Fix #1218

fatg3erman opened the pull request:
Output the last_modified timestamp from mopidy's track model to mpd clients
in the same format as mpd uses - yyyy-mm-ddTHH:MM:SS

Outputs nothing for Last-Modified if last_modified is None

fatg3erman commented:
I don't understand why this is failing. The failure is occurring in code I haven't changed, and it is all passing when I run tox locally.

adamcik commented:
87bf261345919e90cb88853165fb1556046c80ef fixed this, turns out we've had this unnoticed typo for a few years...

jodal commented:
Does MPD include a trailing 'Z' or similar to signal the time zone? Since we are converting from Unix epoch, we are time zone aware. It would be nice if that was reflected in the timestamp to the MPD client.

fatg3erman commented:
Yes it does, at least for me MPD returns a trailing Z. Now I need to figure out how to get python to do the same. Any hints..? :)

jodal commented on tests/mpd/test_translator.py:
```diff
@@ -79,6 +80,31 @@ def test_track_to_mpd_format_for_nonempty_track(self):
+        self.assertIn(('Track', '7/13'), result)
+        self.assertIn(('Date', '1977-01-01'), result)
+        self.assertIn(('Disc', 1), result)
+        self.assertIn(('Last-Modified', datestring), result)
```
I think datestring should be written out here, instead of calculating it with the same logic as used in the implementation.

jodal commented on mopidy/mpd/translator.py:
```diff
@@ -86,6 +87,11 @@ def track_to_mpd_format(track, position=None, stream_title=None):
         result.append(('Disc', track.disc_no))
 
+    if track.last_modified is not None:
+        datestring = datetime.datetime.fromtimestamp(
```
Use `datetime.datetime.utcfromtimestamp` instead, and this will no longer depend on the local time zone.

Then you can safely just append "Z" to the returned string yourself, since the returned timestamp will always be in the UTC time zone.

fatg3erman commented:
OK I've updates this to use UTC time, and changed the test to use a predefined string.
Also changed it from 'if lastmodified is not None' to just 'if lastmodified', as I think I recall once seeing something having a zero for lastmodified, and that's not helpful or useful to a client.

jodal commented:
Haven't looked at the code, but the 0 case sounds like something we want a
test for.

On Tue, Jul 21, 2015, 19:25 fatg3erman <notifications@github.com> wrote:

> OK I've updates this to use UTC time, and changed the test to use a
> predefined string.
> Also changed it from 'if lastmodified is not None' to just 'if
> lastmodified', as I think I recall once seeing something having a zero for
> lastmodified, and that's not helpful or useful to a client.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mopidy/mopidy/pull/1219#issuecomment-123408807>.
>

fatg3erman commented:
OK, added a test for the case when last_modified is 0.

jodal commented:
I merged and simplified the tests to not repeat too much from the previous test in commit 6cb48f29ceeb74a030c4f8086c9e1ded085bd925.
